libgalois/include/katana/PropertyGraph.h

```diff
@@ -196,23 +196,39 @@ class KATANA_EXPORT PropertyGraph {
   static Result<std::unique_ptr<PropertyGraph>> Make(
       std::unique_ptr<tsuba::RDGFile> rdg_file, tsuba::RDG&& rdg);
 
-  /// Make a property graph from an RDG name.
+  /// Make a property graph from an RDG name. Load the default partition and all
+  /// properties.
   static Result<std::unique_ptr<PropertyGraph>> Make(
       const std::string& rdg_name);
 
-  /// Make a property graph from an RDG but only load the named node and edge
-  /// properties.
-  ///
-  /// The order of properties in the resulting graph will match the order of
-  /// given in the property arguments.
+  /// Make a property graph from an RDG name. Load the requested partition and
+  /// all properties.
   ///
-  /// \returns invalid_argument if any property is not found or if there
-  /// are multiple properties with the same name
+  /// This is probably not the Make you are looking for. This is used by the
+  /// partitioner, which makes its living manipulating graph representation
+  /// internals.
+  static Result<std::unique_ptr<PropertyGraph>> Make(
+      const std::string& rdg_name, uint32_t host_to_load);
+
+  /// Make a property graph from an RDG name. Load the default partition and the
+  /// named node and edge properties.
   static Result<std::unique_ptr<PropertyGraph>> Make(
       const std::string& rdg_name,
       const std::vector<std::string>& node_properties,
       const std::vector<std::string>& edge_properties);
 
+  /// Full strength PropertyGraph::Make. Make a property graph from an RDG name.
+  /// Optionally specify a partition to load. Optionally specify properties to
+  /// load.
+  ///
+  /// This is probably not the Make you are looking for. This is used by the
+  /// partitioner, which makes its living manipulating graph representation
+  /// internals.
+  static Result<std::unique_ptr<PropertyGraph>> Make(
+      const std::string& rdg_name, std::optional<uint32_t> host_to_load,
+      const std::vector<std::string>* node_properties,
```

Comment:
Just confirming `std::optional<std::vector<std::string>>` is not satisfactory because of the mandatory initial copy and the downstream `std::move` s required to reduce copies?

Anyway, can you add a comment about if these are `nullptr`, this means load all properties?

Reply:
Besides excessive copying, the big reason I used a `const` pointer was that `RDG` already used `const` pointers and `PropertyGraph` used references, so switching both to optional vectors seemed silly.

I will add that comment.

Reply:
I think we should either go for a single function with optional/default arguments or go for multiple overloaded functions with no optional/default arguments. 

As `PropertyGraph` is more user-facing or high-level, 4 `Make` functions is good but I don't understand why this last one requires optional/pointer (nullptr) arguments. It's better to make the last function:
```cpp
static Result<std::unique_ptr<PropertyGraph>> Make(
      const std::string& rdg_name, uint32_t host_to_load,
      const std::vector<std::string>& node_properties,
      const std::vector<std::string>& edge_properties);
```
Does this have any disadvantages that I'm missing?

As `RDG` is more intra-library or low-level, 1 `Make` is better than the current 2:
```cpp
  static katana::Result<RDG> Make(
      RDGHandle handle, std::optional<uint32_t> host_to_load = std::nullopt,
      const std::vector<std::string>* node_props = nullptr,
      const std::vector<std::string>* edge_props = nullptr);
```
It's okay to require the users of this function to specify `host_to_load` if they need to selectively load node/edge properties.

Reply:
The goal of the last `PropertyGraph::Make` is to be a catch-all that defines the full range of possible behaviors of the function, but I agree that it is a little awkward to not complete the `host_to_load` X `props` cross product. The real benefit of this version is that it allows `Distribution::Make` in the enterprise repo to collect a properly formed set of parameters in one place and then call `PropertyGraph::Make` from there. If `PropertyGraph` only has `Make`s with non-optional parameters then each of the `Distribution::Make`s with non-optional parameters has to call a separate `PropertyGraph::Make`.

So I would be on board with adding the `PG::Make` you described, but I would also like to keep the one that is in there currently. As for `RDG::Make`, I am happy to drop down to one. My main concern was the one you addressed about requiring callers to specify `host_to_load` when they want the default value.

Reply:
If you want a catch-all, then only have the catch-all. We should avoid providing 2 functions/ways to do the same thing. On the other hand, if there is no catch-all, it's fine if each `Distribution::Make` has to call a different `PropertyGraph::Make`. 

Is there any strong arguments against this: either a single function with optional/default arguments or multiple overloaded functions with no optional/default arguments?

Reply:
1) My issue with a single function is that it has more complicated semantics and puts the burden on callers to understand them.
2) My issue with multiple overloaded functions is what I mentioned above about having to call different functions from different overloads lower in the stack. I think overloads are nice from  a caller's perspective, but I dislike replicating logic across the overloads.
3) I also see the appeal of choosing one approach, I think it could be confusing to offer both even if they are clearly delineated in comments. Is there another reason to choose only one that I missed?

Obviously we can only have two of the above three options and it sounds like 3 is non-negotiable for you. In that case I would probably prefer to have 3 and 2 and drop the `Make` that offers optional parameters in `PropertyGraph`.

Reply:
Well summarized. It is negotiable. I just have a strong preference for 1 or 2. 

What do @ddn0 and @tylershunt think about this?

Reply:
Small point: I dislike the mechanism of default arguments because they are based on position and have funny interactions with virtual functions (the latter is not applicable here). Config classes are usually better options get names and you have more freedom to rearrange and grow without breaking callers.

It seems like this issue about hosts and properties in `PropertyGraph::Make` arises because `PropertyGraph` (not host aware) constructs an `RDG` (very host aware). Is it possible to just construct an `RDG` and then give it to `PropertyGraph::Make` instead?

Reply:
I would like to avoid forcing callers to construct RDGs before calling PropertyGraph::Make. I am starting to think that we should just bite the bullet and create a config class.

Reply:
I like the idea of having a config class instead of option 1. It's generable, extensible, and doesn't add that more burden to users IMO.

Reply:
> I would like to avoid forcing callers to construct RDGs before calling PropertyGraph::Make

FWIW, I was thinking that the host-aware construction (katana-users) would go through `PropertyGraph::Make(RDG)` but the host-ambivalent construction (general single machine application users) could still go through `PropertyGraph::Make(string, ....)`